add support for running dependency resolution in parallel

abhi-sanoujam opened the pull request:
Added support to run dependency resolution in parrallel. Can be disabled by passing -Ddependency.version.check.useParallelDepResolution=false in mvn command line (true by default)

Running on presto-main at presto project gives about 4x perf improvement
Without parallel dependency resolution:

```
mvn validate -pl presto-main -Dbuildtime.output.log=true  -Ddependency.version.check.useParallelDepResolution=false
...
[INFO] --- maven-dependency-versions-check-plugin:2.0.3-SNAPSHOT:check (default) @ presto-main ---
[INFO] Checking dependency versions
[INFO] Using parallel dependency resolution: false
...
[INFO]   maven-dependency-versions-check-plugin:check (default) ... [8.796s]
```

Running with parallel resolution:

```
mvn validate -pl presto-main -Dbuildtime.output.log=true 
...
[INFO] --- maven-dependency-versions-check-plugin:2.0.3-SNAPSHOT:check (default) @ presto-main ---
[INFO] Checking dependency versions
[INFO] Using parallel dependency resolution: true
...
[INFO]   maven-dependency-versions-check-plugin:check (default) ... [2.100s]
```

Commit:
add support for running dependency resolution in parallel

electrum commented:
Nice work! Two initial comments:
1. Start commit message titles with an uppercase. We follow this format: http://chris.beams.io/posts/git-commit/
2. The pull request should be created to merge into the original repository (instead of being a pull request for the fork).

electrum commented on src/main/java/com/ning/maven/plugins/dependencyversionscheck/AbstractDependencyVersionsMojo.java:
```diff
@@ -241,6 +245,18 @@
     /** Keeps track of the longest name for an artifact for printing out nicely. */
     protected int maxLen = -1;
 
+    private final ExecutorService executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 5, new ThreadFactory() {
```
Use `ThreadFactoryBuilder` from Guava:

``` java
private final ExecutorService executorService = Executors.newFixedThreadPool(
        Runtime.getRuntime().availableProcessors() * 5,
        new ThreadFactoryBuilder()
                .setNameFormat("dependency-version-check-worker-%s")
                .setDaemon(true)
                .build());
```

abhi-sanoujam replied:
Guava is not available as a dependency - so can't use ThreadFactoryBuilder

electrum replied:
You can add it (this plugin is old enough to be originally written before Guava existed)

electrum commented on src/main/java/com/ning/maven/plugins/dependencyversionscheck/AbstractDependencyVersionsMojo.java:
```diff
@@ -241,6 +245,18 @@
     /** Keeps track of the longest name for an artifact for printing out nicely. */
     protected int maxLen = -1;
 
+    private final ExecutorService executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 5, new ThreadFactory() {
```
I think it's fine to have the threads be daemon threads just to be safe, but pool should be shutdown at the end of the `execute()` method, as the JVM running Maven could be around for a long time.

Regarding the object lifecycle of a `Mojo`, I'm having trouble finding explicit documentation on it, but based on how they are used, it seems the only reasonable way is that they are created once per execution (so `execute()` would be called exactly once).

Commit:
Add guava dependency and use guava utility

Use ThreadFactoryBuilder, Striped instead of locking on string intern
Use Mojo param instead of system prop

Commit:
make main thread await on results from worker threads

electrum commented on src/main/java/com/ning/maven/plugins/dependencyversionscheck/AbstractDependencyVersionsMojo.java:
```diff
@@ -407,66 +423,96 @@ protected Map buildResolutionMap(final String scope) throws MojoExecutionExcepti
         // Map from artifactName --> list of resolutions found on the tree
         final SortedMap resolutionMap = new TreeMap();
+        final List errors = new ArrayList();
+        boolean useParallelDepResolution = "true".equalsIgnoreCase(getPropertyOrDefault("dependency.version.check.useParallelDepResolution", "true"));
```
This should be done as a normal Mojo parameter, so it can be configured in the POM. We can also give it a property as part of the Mojo parameter, so that it can be overridden on the Maven command line with `-D`.

http://books.sonatype.com/mvnref-book/reference/writing-plugins-sect-mojo-params.html

electrum commented on src/main/java/com/ning/maven/plugins/dependencyversionscheck/AbstractDependencyVersionsMojo.java:
```diff
@@ -407,66 +424,106 @@ protected Map buildResolutionMap(final String scope) throws MojoExecutionExcepti
+        final List errors = new ArrayList();
+        final CountDownLatch latch = new CountDownLatch(project.getDependencies().size());
+        boolean useParallelDepResolution = "true".equalsIgnoreCase(getPropertyOrDefault("dependency.version.check.useParallelDepResolution", "true"));
+        LOG.info("Using parallel dependency resolution: " + useParallelDepResolution);
```
This should be debug level. We don't want it to print every time.

electrum commented on src/main/java/com/ning/maven/plugins/dependencyversionscheck/AbstractDependencyVersionsMojo.java:
```diff
@@ -487,23 +544,25 @@ private boolean isVisible(final String scope, final String [] validScopes)
-            if (!existingResolution.getActualVersion().equals(resolution.getActualVersion())) {
-                LOG.warn("Dependency '{} expects version '{}' but '{}' already resolved to '{}'!", 
-                         new Object [] { resolution.getDependencyName(), resolution.getActualVersion(), existingResolution.getDependencyName(), existingResolution.getActualVersion() });
+        synchronized (resolution.getDependencyName().intern()) {
```
This kind of locking is bad practice. The string object is global, so for example, if multiple instances of this plugin is running (which is the case for a parallel build), they will block each other. You could also have something unrelated in the same JVM messing with the same lock.

Guava `Cache` can be used here to replace the map: http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/cache/Cache.html#get(K,%20java.util.concurrent.Callable)

It has a per-key lock around the loading method.

https://github.com/google/guava/wiki/CachesExplained

electrum replied:
Note that `ConcurrentHashMap` in Java 8 has `computeIfAbsent` which has similar semantics (though the locks are larger, so Guava cache should be used if the compute method is not fast).

Commit:
Merge branch 'perf' of https://github.com/abhi-sanoujam/maven-dependency-versions-check-plugin into perf

Commit:
Fix review comments

- Use Futures.allAsList for simpler exception handling
- Use shutdownNow instead of shutdown
- Restore interrupt status
- Use synchronized map
